### PR TITLE
Feature/509 dkg v3

### LINKED ISSUE #509: add BITE2-compatible DKG support
the proposal is described here https://github.com/skalenetwork/internal-support/blob/main/docs/specifications/sgx/reencryption.md

need to update existing DKG implementation to follow the updated procedure:
- add `trustedDecryptDkgSecretV3` and corresponding `generateDKGPolyV3` RPC API method that uses previous bls secret key as a free-term and generates other terms randomly. new RPC method has one additional parameter - previous BLS key name and executes different enclave's function. new enclave call receives the same input parameters and encrypted previous BLS key  
- add `trustedCreateBlsKeyV3` and corresponding `createBLSPrivateKeyV3` RPC API method that generates bls key share from the subset of `t` out of `n` nodes (3 out of 4, 11 out of 16, etc). new methods receive `t` instead of `n` shares to construct BLS key and corresponding indexes, and index of the node a key is generated for

### PR DESCRIPTION
**Pull request by PropzSaladaz**
## Description

This PR implements DKG V3 key rotation for SGX wallet. The main goal is to support rotating DKG/BLS key shares while **preserving the same common BLS public key**, so ciphertexts encrypted before rotation can still be decrypted by the new committee as long as the threshold assumptions hold.

The core requirement added by this PR is that a new DKG polynomial can be generated from a previous BLS key share. Instead of choosing a fully random polynomial, DKG V3 anchors the polynomial free coefficient to the node’s previous BLS private key. This lets the new set of generated shares reconstruct the same common secret/public key while changing the individual node shares.

## Enclave Changes

Two new enclave ECALLs were added:

### `trustedGenDkgSecretV3`

This method generates a DKG polynomial whose free coefficient is the previous BLS private key.

Behavior:
- Decrypts the previous BLS key inside the enclave.
- Generates a DKG polynomial using the previous BLS scalar as coefficient `0`.
- Randomly generates the remaining coefficients.

This is the main primitive that preserves the old common BLS public key across rotation.

### `trustedCreateBlsKeyV3`

This method creates a new BLS private key share **from a subset** of encrypted secret contributions.

Behavior:
- Computes Lagrange coefficients for the provided contributor indices.
- Decrypts each encrypted secret contribution using the existing ECDH/session-key flow.
- Weights each contribution by its Lagrange coefficient.
- Encrypts the resulting BLS private key as a non-exportable BLS key.

The contributor indices are explicit because DKG V3 may build a key from only a threshold subset of contributors, and those contributors may not be contiguous.

## Server / Crypto Layer

`SGXWalletServer` now exposes two new internal implementations:

### `generateDKGPolyV3Impl(polyName, previousBLSPrivateKeyName, t)`

This validates:
- `polyName` must be a valid `POLY` name.
- `previousBLSPrivateKeyName` must be a valid `BLS_KEY` name.
- `t` must be in the accepted range.

It then reads the previous encrypted BLS key from LevelDB, calls the enclave V3 polynomial generation flow, and stores the encrypted polynomial under `polyName`.

### `createBLSPrivateKeyV3Impl(blsKeyName, ethKeyName, polyName, secretContributions, t, n)`

This validates:
- recipient ECDSA key name
- output BLS key name
- optional polynomial name
- DKG `(t, n)` parameters
- `secretContributions` is an array
- contribution count is between `t` and `n`
- contributor indices are unique and in range
- each secret share is the expected 96-byte hex value

It then reads the recipient ECDSA key, calls `createBLSShareV3`, stores the new BLS key, and optionally cleans up temporary DKG data if `polyName` is provided.

**`polyName` is optional for `createBLSPrivateKeyV3` because joining nodes may not have generated a local polynomial and therefore have no local polynomial data to clean up.**

### JSONRPC API

Two JSONRPC methods were added.

`generateDKGPolyV3`

Params:
- `polyName`
- `previousBLSPrivateKeyName`
- `t`

This creates a new DKG V3 polynomial using the previous BLS key as the polynomial free coefficient.

`createBLSPrivateKeyV3`

Params:
- `blsKeyName`
- `ethKeyName`
- optional `polyName`
- `secretContributions`
- `t`
- `n`

`secretContributions` is an array of objects:

```json
{
  "contributorIndex": 0,
  "secretShare": "192_hex_chars"
}
```

This creates the recipient’s new BLS key share using the provided encrypted contributions and their contributor indices.

The JSONRPC spec was updated, and `StubClient`/`AbstractStubServer` were updated manually. The manual edits are intentional for this PR: **the current JSONRPC auto-generator is not producing compilable output after regeneration. Fixing the generator or generated-code workflow should be handled in a separate PR.**

### ZMQ API

The same DKG V3 APIs were added to ZMQ.

New request/response message types:
- `generateDKGPolyV3Req`
- `generateDKGPolyV3Rsp`
- `createBLSPrivateV3Req`
- `createBLSPrivateV3Rsp`

`ZMQClient` now supports:
- `generateDKGPolyV3(polyName, previousBLSPrivateKeyName, t)`
- `createBLSPrivateKeyV3(blsKeyName, ethKeyName, polyName, secretContributions, t, n)`

ZMQ request processing calls the same `SGXWalletServer` implementation as JSONRPC, so validation and behavior are shared.

## Tests Added

This PR adds tests at three levels.

### Enclave-level DKG V3 tests:
- Verifies `trustedGenDkgSecretV3` uses the previous BLS key as polynomial coefficient `0`.
- Verifies the first public share from the generated polynomial matches the public key of the previous BLS key.
- Verifies `trustedCreateBlsKeyV3` can create a BLS key from encrypted secret contributions.

### Rotation/liveness tests:
- Runs DKG V2, encrypts threshold-encryption payloads, rotates to DKG V3, and verifies DKG V3 keys can decrypt old ciphertexts.
- Verifies the common BLS public key remains stable across rotation.
- Tests rotation with the same committee.
- Tests rotation with joining nodes.
- Tests non-responding node scenarios where decryption succeeds if enough shares remain and fails when fewer than `t` shares are available.

### Security tests:
- Verifies mixed V2/V3 shares fail to decrypt.
- Demonstrates why rotating out a full threshold of old nodes is unsafe: retired threshold nodes can still decrypt old ciphertexts.
- Verifies retired nodes from separate rotations cannot collude if each rotation retired fewer than `t` nodes, even if the total retired node count across epochs exceeds `t`.
- Tests boundary cases around rotated and non-responding nodes.

### API tests:
- JSONRPC tests for `generateDKGPolyV3`.
- JSONRPC tests for `createBLSPrivateKeyV3`.
- ZMQ tests for `generateDKGPolyV3`.
- ZMQ tests for `createBLSPrivateKeyV3`.
- Negative coverage for invalid polynomial names, invalid previous BLS names, invalid thresholds, malformed contribution payloads, invalid key names, invalid optional `polyName`, and too few contributions.

`testw.py` was updated to include the new DKG V3 enclave, API, rotation, and security test tags.

### CODE REVIEW

File: SGXWalletServer.cpp [hunk 1]
```diff
@@ -1057,6 +1118,166 @@ Json::Value SGXWalletServer::createBLSPrivateKeyV2Impl(
...
+                               "], missing required field: contributorIndex");
+      }
+      if (!entry["contributorIndex"].isUInt()) {
+        throw SGXException(INVALID_CREATE_BLS_KEY_SECRET_SHARES_LENGTH,
+                           string(__FUNCTION__) +
+                               ":Invalid secretContributions[" + to_string(i) +
+                               "].contributorIndex, expected: unsigned "
+                               "integer, got: " +
+                               jsonTypeName(entry["contributorIndex"]));
+      }
```
[the remark this answers is not in the document]
Comment (PropzSaladaz) [reply SGXWalletServer.cpp:1196]: Negative values are intentionally invalid here: `contributorIndex` is a participant index in `[0, n - 1]`, not an arbitrary field x-coordinate. Also, `JSONcpp isUInt()` already accepts non-negative intValues, so `entry["contributorIndex"] = contributor` does not fail as long as `contributor >= 0`. No change to accept negative or general signed ints is needed.

File: secure_enclave/DHDkg.c [hunk 2]
```diff
@@ -79,9 +79,14 @@ int gen_session_key(char *skey_str, char *pb_keyB, char *common_key) {
...
+        goto clean;
+    }
 
-    point_set_hex(pub_keyB, pb_keyB_x, pb_keyB_y);
+    if (point_set_hex(pub_keyB, pb_keyB_x, pb_keyB_y) != 0) {
+        LOG_ERROR("gen_session_key: invalid public key point");
+        goto clean;
+    }
 
     point_multiplication(session_key, skey, pub_keyB, curve);
```
[the remark this answers is not in the document]
Comment (PropzSaladaz) [reply secure_enclave/DHDkg.c:91]: fixed

File: secure_enclave/DHDkg.c [hunk 3]
```diff
@@ -139,11 +138,21 @@ int session_key_recover(const char *skey_str, const char *sshare, char *common_k
...
         goto clean;
     }
 
-    point_set_hex(pub_keyB, pb_keyB_x, pb_keyB_y);
+    if (point_set_hex(pub_keyB, pb_keyB_x, pb_keyB_y) != 0) {
+        LOG_ERROR("session_key_recover: invalid public key point");
+        goto clean;
+    }
+
     point_multiplication(session_key, skey, pub_keyB, curve);
```
[the remark this answers is not in the document]
Comment (PropzSaladaz) [reply secure_enclave/DHDkg.c:156]: fixed

**Review by kladkogex — APPROVED** [review]

**Review by badrogger — APPROVED** [review]
